### ISSUE TITLE
Rebuild Epoch Parsing

### ISSUE THREAD

**Issue by superstructor** [day8/re-frame-10x#302]
The epoch parsing code is in need of significant refactoring and bug fixes. It has led to the following issues:

- delays in rendering content: #264
- lagging event history: #268
- `dispatch-sync` invisible: #154
- `nil` app-db state: #263, #151
- inconsistent use of event filters: #175
- duplication of subs between the current epoch and intra-epoch subs
- invalid timing information

The relevant code is in [`::receive-new-traces`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L16), [`parse-traces`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/tools/metamorphic.cljc#L219), surrounding code in the `metamorphic` ns and in subs that use the app-db state derived from the epoch parsing.

## Step 1: Move `ignored-events` filtering from event to subs

In `::recieve-new-traces` there is [a `remove` to filter ignored events](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L30) (via settings) before epochs are stored in app-db.

This causes some issues:
1. Partial epoch states in app-db leading to [odd workarounds](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/subs.cljs#L54) in subscriptions which are not all handled
2. If the user removes or adds an epoch filter with existing traces in app-db, 10x is in an invalid state as it does not have a correct 'view' of epochs according to the settings.

The `remove` filter needs to be deleted from `::recieve-new-traces` and added to the appropriate subs instead.

## Step 2: Change `metamorphic/parse-traces` so that `dispatch-sync` is its own epoch

The delimitation of `dispatch-sync` epochs is not detected in [`metamorphic/parse-traces`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/tools/metamorphic.cljc#L219). This needs to be fixed so a `dispatch-sync` is returned as a distinct epoch.

## Step 3: Move all parsing from `::recieve-new-traces` event to `metamorphic` ns
 
 [`::receive-new-traces`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L16) contains too much data transformation/parsing related code. If this is moved to `metamorphic` ns it would be easier to write unit tests to debug/check correct behavior and so forth.

New APIs/functions should be considered as part of this move, as should renaming. The current names are certainly not great.

Tests should be added for all the significant epoch parsing use cases/situations.

At least the following should be moved: 
- [`parse-traces`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L26)
- [`parse-state`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L29) changes
- [`subscription-info`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L38)
- [`subscription-match-state`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L40)
- [`subscription-matches`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L41)
- [`new-sub-state`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L43)
- [`timing`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L44) mapv
- [`new-matches`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L63) map
- [`quiescent?`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L70)
- [`all-matches`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L72)
- [`retained-matches`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L73)
- [`first-id-to-retain`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L74)
- [`retained-traces`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L75)
- [`:matches-by-id`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L86) into...map
- [`:selected-epoch-id`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L92)
- [`:selected-epoch-index`](https://github.com/day8/re-frame-10x/blob/1dff52ff6024b144d259a1a091564cab82ba9fec/src/day8/re_frame_10x/navigation/epochs/events.cljs#L93)

## Step 4: Performance 

Epoch parsing and associated code is currently the biggest hit on 10x performance.

Once more of the code is moved into functions in `metamorphic` with unit tests various optimisations could be explored such as storing the internal parse state as transient vs persistent collections, doing less string parsing etc.

**Comment by mike-thompson-day8** [comment]
Closing — superseded by #424 ("Adopt re-frame's native register-epoch-cb as the epoch source; retire metamorphic/parse-traces").

Status as of close:

- All 6 motivating sub-bugs (#154, #263, #151, #175, #264, #268) have been closed individually since 2021.
- The 4 refactor steps in this issue's body remain mostly unimplemented (Step 1: filter-in-event still inline; Step 3: `::receive-new-traces` is still a 70-line let-cascade; Step 4: no transient/perf work).
- But the right move now isn't to refactor the trace-stream parser — it's to retire it. re-frame upstream now ships `register-epoch-cb` (rf-ybv), and re-frame-pair has been consuming it in production. 10x can swap the parser for a thin adapter on the native callback, with the existing parser kept as the fallback for older re-frame.

#424 captures that work end-to-end (motivation, approach, fields-to-preserve audit, backwards-compat strategy, acceptance criteria, suggested ordering). Once #424 lands, the architectural fragility this issue named is structurally gone.
